src/core/thread/mle_router.cpp

```diff
@@ -1108,20 +1108,32 @@ ThreadError MleRouter::HandleAdvertisement(const Message &aMessage, const Ip6::M
                      leaderData.GetWeighting(), peerParitionId,
                      mLeaderData.GetWeighting(), mLeaderData.GetPartitionId());
 
-        if ((leaderData.GetWeighting() > mLeaderData.GetWeighting()) ||
-            (leaderData.GetWeighting() == mLeaderData.GetWeighting() &&
-             peerParitionId > mLeaderData.GetPartitionId()))
+        if (mDeviceMode & ModeTlv::kModeFFD)
         {
-            otLogDebgMle("trying to migrate\n");
-            BecomeChild(kMleAttachBetterPartition);
+            if ((leaderData.GetWeighting() > mLeaderData.GetWeighting()) ||
+                (leaderData.GetWeighting() == mLeaderData.GetWeighting() &&
+                 peerParitionId > mLeaderData.GetPartitionId()))
+            {
+                otLogDebgMle("trying to migrate\n");
+                BecomeChild(kMleAttachBetterPartition);
+            }
+            else
+            {
+                error = kThreadError_Drop;
+            }
         }
 
-        ExitNow(error = kThreadError_Drop);
+        ExitNow();
     }
     else if (leaderData.GetLeaderRouterId() != GetLeaderId())
     {
-        BecomeDetached();
-        ExitNow(error = kThreadError_Drop);
+        if (mDeviceMode & ModeTlv::kModeFFD)
+        {
+            BecomeDetached();
```

Comment:
Should this only happen if the device is currently a router?  If the device is a child, it should just follow it's parent, right?